### PR TITLE
Feat(autocomplete): listbox virtualization to increase performance (6s -> 0s)

### LINKED ISSUE #3047: [Feature Request] Optimize AutoCompelete and Select components for large options.
### Is your feature request related to a problem? Please describe.

First of all, nice library. I would say I have been following this library before its birth. I love the way `Vuesax` looks, so I was looking for something similar in the React world. You came at the right time. Nice work once again!

I have about `270,000+` options while using the `Autocomplete` component. Before you call me crazy, I'm only fetching the data when the user inputs a search query. Again, I'm not crazy. I'm using debounce from `Lodash`, and every good stuff that will make it work right. But, it is too slow. i mean very slow.

### Describe the solution you'd like

I would suggest making it efficient like `react-select`, I tried using it, but I did not get how to merge both. I wanted to have the functionalities of it + `next UI` beautiful Autocomplete component.

I don't mind if someone could provide an example

### Describe alternatives you've considered

`react-select`

### Screenshots or Videos

_No response_

### PR DESCRIPTION
**Pull request by vinroger**
Closes #3047 

## 📝 Description
This PR adds virtualization support to the Autocomplete component to improve performance when handling large datasets. The implementation uses the `react-window` library to efficiently render only the visible portion of the list

## ⛳️ Current behavior (updates)
Previously, the Autocomplete component rendered all options in the DOM, which led to performance issues if we have large lists. 

The delays in rendering and interacting with the component reach about 3-5 seconds for 1,000 lists.

Upon profiling, I discovered that the delay in the autocomplete feature is primarily due to the rendering of the DOM for the React component list, as there is currently no virtualization implemented.

**Image**: Delay in the multiple react dom function call (creating 1000 list items) -> amounting to total ~26 seconds for 1000 items

image

**Image**: Delay in the actual search algorithm (or finding the matched items), which takes not-so-significant time (approximately 2-3 ms).

image

Details about benchmark on my comment on the github issue: https://github.com/nextui-org/nextui/issues/3047#issuecomment-2452957301

**Question: Why not using `virtualizer` by react-aria?** https://github.com/nextui-org/nextui/issues/3047#issuecomment-2453860455
Based on RA release note on July 2024, the virtualizer component is "unstable" and they mentioned that it is not intended for public use.

Link to RA release notes: https://react-spectrum.adobe.com/releases/2024-07-22.html

"We are also working on bringing virtualization support to React Aria Components, which is available as an unstable API in this version."

image

There is no mention of virtualizer update ever since. July 2024 was their second latest release.
Hence, I proceed with using react-window, https://www.npmjs.com/package/react-window

## 🚀 New behavior

With the introduction of virtualization using react-window, the Autocomplete component now renders only the list items that are visible within the scroll viewport, plus a small buffer. 

This will enhance performance, making the component responsive even with large datasets.

## 💣 Is this a breaking change (Yes/No): No.

This change is not breaking. The existing API remains the same, existing users can continue using the component as before, while benefiting from improved performance.

## 📝 Additional Information

The virtualized list is configured with a `maxListboxHeight` and `itemHeight` to ensure consistent rendering performance.

Performance tests show significant improvements in rendering time and responsiveness. (from 5-10s to unnnoticeable)
To test: autocomplete with large datasets (>10,000 items).

## Summary by CodeRabbit

- **New Features**
	- Introduced a `VirtualizedListbox` component for improved performance in the `Autocomplete` feature.
	- Added support for handling large datasets in the `Autocomplete` component with new templates for testing.
  
- **Bug Fixes**
	- Removed maximum height constraint from the `listboxWrapper` style for better adaptability.

- **Documentation**
	- Updated type definitions to include new properties for virtualization in the autocomplete functionality. 

- **Chores**
	- Added a new dependency for virtualization support in the listbox component.

### CODE REVIEW

File: packages/components/autocomplete/src/virtualizedlistbox.tsx [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+  return (
+    <div {...getBaseProps()}>
+      {topContent}
+      {content}
+      {bottomContent}
+    </div>
+  );
+}
+
+VirtualizedListbox.displayName = "NextUI.Listbox";
```
[the remark this answers is not in the document]
Comment (jrgarciadev) [reply packages/components/autocomplete/src/virtualizedlistbox.tsx:113]: Move `VirtualizedListbox` component to the `listbox` component package

**Comment by jrgarciadev** [comment]
- Add tests
- Add a storybook

**Comment by ryo-manba** [comment]
Thanks for the PR. This issue might be resolved by using the new [virtualizer from react-aria](https://react-spectrum.adobe.com/releases/2024-07-22.html#api-updates). I'll start working on this approach. I believe it could be a better solution than adding new dependencies.

**Comment by vinroger** [comment]
Hi @ryo-manba, in the react-aria docs, the virtualizer component is still marked as "unstable" (in their storybook stories) and "not intended for public use" (in the release notes)

> Because of various breaking changes, several packages are being bumped a major version, specifically @react-aria/virtualizer , @react-stately/virtualizer , and @react-stately/layout. While these packages were undocumented and not yet intended for public use, we know this is not always the case, so please review the updated code carefully.

What do you think of this?
Also if we decided to move forward with the virtualizer from react-aria, may I take this feature? :) 🙇‍♂️🙇‍♂️

Thanks!

cc: @jrgarciadev

**Comment by ryo-manba** [comment]
Sorry, I missed that this was already mentioned in the PR.
Thanks for pointing it out! Please go ahead as planned.

**Comment by vinroger** [comment]
Hi! @jrgarciadev 

Thanks a lot for the feedbacks - I have updated the PR according to your requests. 

### Tests / Storybook Stories
- Added tests for autocomplete with large dataset (1,000 and 10,000 dataset)
- Added tests for autocomplete with custom itemHeight and maxListboxHeight (passed as props to autocomplete components)

### Listbox
- Implemented isVirtualized logic in virtualized-listbox.tsx
- Moved virtualized-listbox to the listbox components folder
- Implemented maxListboxHeight, itemHeight as props (no longer static value) -> this adds 2 props to the autocomplete component (use-autocomplete.tsx)

### Miscellaneous
- Misc enhancement: fix wrong imports, rename component name, rename props name

Let me know if you have any more change requests!

File: packages/core/theme/src/components/autocomplete.ts [hunk 2]
```diff
@@ -5,7 +5,7 @@ import {tv} from "../utils/tv";
 const autocomplete = tv({
   slots: {
     base: "group inline-flex flex-column w-full",
-    listboxWrapper: "scroll-py-6 max-h-64 w-full",
+    listboxWrapper: "scroll-py-6 w-full",
```
Comment (vinroger) [inline packages/core/theme/src/components/autocomplete.ts:8]: to allow for custom popover (listbox) height

**Comment by jrgarciadev** [comment]
Hey @vinroger Thanks for updating the code! I should have mentioned this sooner, but could you use [TanStack’s React Virtual](https://tanstack.com/virtual/latest/docs/framework/react/react-virtual) instead of react-window? The react-window package hasn’t been updated in years, and I don’t want it to prevent us from fully supporting React 19 and Next.js 15

Also the default behaviour should be `isVirtualized: false`

**Review by jrgarciadev — APPROVED** [review]

**Comment by jrgarciadev** [comment]
Hey @vinroger I'm closing this in favour of this one https://github.com/nextui-org/nextui/pull/4094 ( I had to push some changes, you gonna appear as the core contributor) thank you! 🙏🏻

**Pull request closed without merging**